test/Microsoft.AspNet.Razor.Runtime.Test/TagHelpers/CompleteTagHelperDescriptorComparer.cs

```diff
@@ -23,16 +23,25 @@ bool IEqualityComparer<TagHelperDescriptor>.Equals(TagHelperDescriptor descripto
             return base.Equals(descriptorX, descriptorY) &&
                    // Tests should be exact casing
                    string.Equals(descriptorX.TagName, descriptorY.TagName, StringComparison.Ordinal) &&
-                   descriptorX.Attributes.SequenceEqual(descriptorY.Attributes,
-                                                        CompleteTagHelperAttributeDescriptorComparer.Default);
+                   string.Equals(descriptorX.Prefix, descriptorY.Prefix, StringComparison.Ordinal) &&
+                   Enumerable.SequenceEqual(
+                       descriptorX.RequiredAttributes,
+                       descriptorY.RequiredAttributes,
+                       StringComparer.OrdinalIgnoreCase) &&
+                   descriptorX.Attributes.SequenceEqual(
+                       descriptorY.Attributes,
+                        CompleteTagHelperAttributeDescriptorComparer.Default);
         }
 
         int IEqualityComparer<TagHelperDescriptor>.GetHashCode(TagHelperDescriptor descriptor)
         {
-            return HashCodeCombiner.Start()
-                                   .Add(base.GetHashCode())
-                                   .Add(descriptor.Attributes)
-                                   .CombinedHash;
+            return HashCodeCombiner
+                .Start()
+                .Add(base.GetHashCode())
+                .Add(descriptor.TagName, StringComparer.Ordinal)
+                .Add(descriptor.Prefix)
+                .Add(descriptor.Attributes)
```

Comment:
the hash code of a `List<T>` changes for every instance; even if they contain the same entries.  so you need to add the hash codes of the entries -- similar to the base class' handling of `RequiredAttributes`.  this also gives you an opportunity to use the `CompleteTagHelperAttributeDescriptorComparer` to get those entry hash codes.